File: frontend/generate-version.js
#!/usr/bin/env node

/**
 * Generate version file with git commit hash
 * This runs during the build process to capture the current git commit
 *
 * Supports two modes:
 * 1. Local dev: Uses git commands to get version info
 * 2. Docker build: Uses GIT_COMMIT and GIT_DATE environment variables
 */

import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let version;

// Check if running in Docker with build args
if (process.env.GIT_COMMIT && process.env.GIT_DATE) {
  console.log('Using git info from environment variables (Docker build)');
  version = {
    commit: process.env.GIT_COMMIT,
    date: process.env.GIT_DATE,
    buildTime: new Date().toISOString()
  };
} else {
  // Try to get git info from git commands (local dev)
  try {
    // Get git commit hash (short 12 chars)
    const gitHash = execSync('git rev-parse --short=12 HEAD')
      .toString()
      .trim();

    // Get git commit date
    const gitDate = execSync('git log -1 --format=%cd --date=short')
      .toString()
      .trim();

    version = {
      commit: gitHash,
      date: gitDate,
      buildTime: new Date().toISOString()
    };

    console.log('✓ Version file generated from git:', version);
  } catch (error) {
    console.warn('⚠ Could not get git info:', error.message);
    // Create fallback version
    version = {
      commit: 'dev',
      date: new Date().toISOString().split('T')[0],
      buildTime: new Date().toISOString()
    };
  }
}

// Write to public folder so it's included in the build
const publicDir = path.join(__dirname, 'public');
if (!fs.existsSync(publicDir)) {
  fs.mkdirSync(publicDir, { recursive: true });
}

fs.writeFileSync(
  path.join(publicDir, 'version.json'),
  JSON.stringify(version, null, 2)
);

console.log('✓ Version file written:', version);

File: generate_version.py
#!/usr/bin/env python3
"""
Generate version.py file with git commit hash
This runs during Docker build or deployment to capture the current git commit
"""

import subprocess
import json
from datetime import datetime
from pathlib import Path

def get_git_info():
    """Get git commit info"""
    try:
        # Get git commit hash (short 12 chars)
        git_hash = subprocess.check_output(
            ['git', 'rev-parse', '--short=12', 'HEAD'],
            stderr=subprocess.DEVNULL
        ).decode('utf-8').strip()

        # Get git commit date
        git_date = subprocess.check_output(
            ['git', 'log', '-1', '--format=%cd', '--date=short'],
            stderr=subprocess.DEVNULL
        ).decode('utf-8').strip()

        return {
            'commit': git_hash,
            'date': git_date,
            'build_time': datetime.utcnow().isoformat() + 'Z'
        }
    except (subprocess.CalledProcessError, FileNotFoundError):
        # Fallback if git is not available or not a git repo
        return {
            'commit': 'dev',
            'date': datetime.utcnow().strftime('%Y-%m-%d'),
            'build_time': datetime.utcnow().isoformat() + 'Z'
        }

def main():
    version_info = get_git_info()

    # Write to app/version.py
    version_file = Path(__file__).parent / 'app' / 'version.py'

    with open(version_file, 'w') as f:
        f.write('"""Auto-generated version information"""\n\n')
        f.write(f"COMMIT = '{version_info['commit']}'\n")
        f.write(f"DATE = '{version_info['date']}'\n")
        f.write(f"BUILD_TIME = '{version_info['build_time']}'\n")
        f.write('\n')
        f.write('def get_version():\n')
        f.write('    """Get version dict"""\n')
        f.write('    return {\n')
        f.write("        'commit': COMMIT,\n")
        f.write("        'date': DATE,\n")
        f.write("        'build_time': BUILD_TIME\n")
        f.write('    }\n')

    print(f'✓ Backend version file generated: {version_info}')

if __name__ == '__main__':
    main()
